Add microbenchmarks for sort operator with reference segment null values and different datatypes

bakoe opened the pull request:
Closes #3

jonasnoki committed:
Extend CostumTableGeneration in sort_benchmark

jonasnoki committed:
Add ReferenceSegment and pico Benchmarks.

jonasnoki committed:
Add String and Null Micro-Benchmarks for Sorting.

jonasnoki commented:
```
Bastian.Koenig@vm-appleton:~/hyrise$ cmake-build-release/hyriseMicroBenchmarks --benchmark_filter=BM_Sort*
2019-12-09 21:29:05
Running cmake-build-release/hyriseMicroBenchmarks
Run on (32 X 2500 MHz CPU s)
CPU Caches:
  L1 Data 32K (x32)
  L1 Instruction 32K (x32)
  L2 Unified 256K (x32)
  L3 Unified 25600K (x32)
Load Average: 0.09, 0.08, 0.03
--------------------------------------------------------------------------------------------
Benchmark                                                  Time             CPU   Iterations
--------------------------------------------------------------------------------------------
SortBenchmark/BM_Sort                                9296478 ns      9296018 ns           90
SortPicoBenchmark/BM_SortPico                           2914 ns         2913 ns       232133
SortSmallBenchmark/BM_SortSmall                       638193 ns       638143 ns         1000
SortLargeBenchmark/BM_SortLarge                     87546992 ns     87537032 ns           10
SortReferenceBenchmark/BM_SortReference             15313964 ns     15313812 ns           50
SortReferencePicoBenchmark/BM_SortReferencePico         3653 ns         3653 ns       192312
SortReferenceSmallBenchmark/BM_SortReferenceSmall     797076 ns       797076 ns         1085
SortReferenceLargeBenchmark/BM_SortReferenceLarge  110309045 ns    110306281 ns            6
SortStringBenchmark/BM_SortString                   25696278 ns     25695991 ns           28
SortStringSmallBenchmark/BM_SortStringSmall          2756916 ns      2756870 ns          319
SortStringLargeBenchmark/BM_SortStringLarge        577616215 ns    577569333 ns            1
SortNullBenchmark/BM_SortNullBenchmark              15189981 ns     15189980 ns           55
```

tobodner commented on src/benchmark/operators/sort_benchmark.cpp:
```diff
@@ -1,22 +1,181 @@
+#include <expression/expression_functional.hpp>
```
This belongs to the (Hyrise) set of header below.

tobodner commented on src/benchmark/operators/sort_benchmark.cpp:
```diff
@@ -1,22 +1,181 @@
 #include "operators/sort.hpp"
 #include "operators/table_wrapper.hpp"
+#include "synthetic_table_generator.hpp"
+#include "operators/join_hash.hpp"
```
Respect the order of headers.

tobodner commented on src/benchmark/operators/sort_benchmark.cpp:
```diff
@@ -1,22 +1,181 @@
-    auto sort = std::make_shared<Sort>(_table_wrapper_a, ColumnID{0} /* "a" */, OrderByMode::Ascending);
-    sort->execute();
+    auto warm_up = std::make_shared<Sort>(_table_wrapper_a, ColumnID{0} /* "a" */, OrderByMode::Ascending);
+    warm_up->execute();
```
I know that this is a best practice. Still wondering: Did you see an actual difference with/out warmup?

jonasnoki replied:
These are the results without the `warm_up`:
```
Bastian.Koenig@vm-appleton:~/hyrise$ cmake-build-release/hyriseMicroBenchmarks --benchmark_filter=BM_Sort*
2019-12-16 17:12:49
Running cmake-build-release/hyriseMicroBenchmarks
Run on (32 X 2500 MHz CPU s)
CPU Caches:
  L1 Data 32K (x32)
  L1 Instruction 32K (x32)
  L2 Unified 256K (x32)
  L3 Unified 25600K (x32)
Load Average: 0.29, 0.14, 0.05
--------------------------------------------------------------------------------------------
Benchmark                                                  Time             CPU   Iterations
--------------------------------------------------------------------------------------------
SortBenchmark/BM_Sort                               10013751 ns     10013638 ns           84
SortPicoBenchmark/BM_SortPico                           3115 ns         3114 ns       213037
SortSmallBenchmark/BM_SortSmall                       701094 ns       701036 ns          746
SortLargeBenchmark/BM_SortLarge                     88207271 ns     88196949 ns            9
SortReferenceBenchmark/BM_SortReference             15162234 ns     15161228 ns           56
SortReferencePicoBenchmark/BM_SortReferencePico         3709 ns         3708 ns       186817
SortReferenceSmallBenchmark/BM_SortReferenceSmall     902671 ns       902670 ns          594
SortReferenceLargeBenchmark/BM_SortReferenceLarge  173210621 ns    173210489 ns            3
SortStringBenchmark/BM_SortString                   32152662 ns     32148606 ns           27
SortStringSmallBenchmark/BM_SortStringSmall          3176690 ns      3176126 ns          179
SortStringLargeBenchmark/BM_SortStringLarge        516888142 ns    516836706 ns            2
SortNullBenchmark/BM_SortNullBenchmark              14963719 ns     14963718 ns           57
```

bakoe replied:
The results show: The microbenchmarks indeed perform noticeably different with and without warm-up – although we're not really sure about which way is “better” (see e.g. the SortNullBenchmark or SortStringLarge, where the warm-up _increased_ the amount of iterations with disabled warm-up instead of _decreasing_ it as with most of the Benchmarks), we think that leaving the warm-up enabled is good at this point.

tobodner commented on src/benchmark/operators/sort_benchmark.cpp:
```diff
@@ -1,22 +1,181 @@
+        row_count, chunk_size, std::vector<SegmentEncodingSpec>(num_columns, {EncodingType::Unencoded}), std::nullopt,
+        UseMvcc::No));
+    _table_wrapper_a->execute();
+  }
```
All in all good, but I am wondering, whether we could reduce redunancy below even more by using the means for benchmark parameterization mentioned in the first PR.

bakoe replied:
:+1: We've created a separate follow-up issue for that (#5) and will look into it latest when we implement the next micro benchmarks involving parameterisation.

tobodner commented on src/benchmark/operators/sort_benchmark.cpp:
```diff
@@ -1,22 +1,181 @@
+
+
+    // TODO(sorting-group): Find a more reliable way to produce an exct number of NULL values. Right now this produces
+    //  ~20% NULL Values of the bigger table
```
IMO, the cleanest way would be to extend SyntheticTableGenerator to allow for a "null ratio" in the produced data.

bakoe replied:
Resolved in 190d705a9b7e052f12796786604a79ea90bf5fe4

bakoe committed:
Merge branch 'master' into feature/3-Microbenchmarks-for-Sort-Operator-with-ReferenceSegment-Null-Values-and-different-Datatypes

# Conflicts:
#	src/benchmark/operators/sort_benchmark.cpp

bakoe committed:
Fix header ordering in sort_benchmark.cpp

bakoe committed:
Fix wrong resolution of merge conflict from master

bakoe committed:
Extend SyntheticTableGenerator to allow NULL

Bouncner commented on src/benchmarklib/synthetic_table_generator.cpp:
```diff
@@ -62,7 +62,8 @@ std::shared_ptr<Table> SyntheticTableGenerator::generate_table(
     const std::optional<ChunkEncodingSpec>& segment_encoding_specs,
-    const std::optional<std::vector<std::string>>& column_names, const UseMvcc use_mvcc) {
+    const std::optional<std::vector<std::string>>& column_names, const UseMvcc use_mvcc,
+    const std::optional<float> null_ratio) {
```
Just short (I just sent a rather concrete email): if you do it that way, please use rather a std::vector with a value for each column. You don't need to use make NULL optional, you could default to 0.0f.

bakoe replied:
@Bouncner Thanks for your very valuable feedback! We created a separate issue (#8) for implementing the proposed changes to the method signature of SyntheticTableGenerator. 🙂

bakoe committed:
Add micro benchmark for multi-column sorting (SQL)

birneamstiel commented on src/benchmarklib/synthetic_table_generator.cpp:
```diff
@@ -149,6 +150,19 @@ std::shared_ptr<Table> SyntheticTableGenerator::generate_table(
 
+          std::vector<bool> null_values;
+
+          if (null_ratio.has_value()) {
```
Add documentation on how we distribute null values here.

bakoe replied:
Resolved in e8b8bb797c3adbc7ed020a820dd14d294087dcc1

bakoe commented on src/benchmark/operators/sort_benchmark.cpp:
```diff
@@ -22,28 +31,173 @@ class SortBenchmark : public MicroBenchmarkBasicFixture {
+        ORDER BY col_1, col_2)";
+
+    for (auto _ : state) {
+      // auto stored_table_node = std::make_shared<StoredTableNode>("table_a");
```
TODO Remove commented-out code here and in line 54

bakoe committed:
Remove commented-out code in sort_benchmark.cpp

bakoe committed:
Add micro benchmark for SQL based single col sort

bakoe committed:
Fix NULL value creation in SyntheticTableGenerator

Previously, the step_size was casted to an integer, making
the loop losing precision for step_size values with a decimal
place, e. g. when using a ratio of 0.75 and thus requiring to
make every 1.33-th element NULL (as opposed to every element).

bakoe committed:
Document SyntheticTableGenerator's NULL generation

bakoe commented on src/benchmark/operators/sort_benchmark.cpp:
```diff
@@ -1,10 +1,19 @@
+#include <SQLParser.h>
+#include <SQLParserResult.h>
+#include <logical_query_plan/stored_table_node.hpp>
```
TODO Fix the `#include` grouping and use `""` vs `<>` appropriately.

bakoe commented on src/benchmark/operators/sort_benchmark.cpp:
```diff
@@ -22,28 +31,197 @@ class SortBenchmark : public MicroBenchmarkBasicFixture {
+    storage_manager.add_table("table_a",
+                              GenerateCustomTable(ChunkOffset{2'000}, size_t{40'000}, DataType::Int, column_names));
+
+    const std::string query = R"(
```
TODO Reduce code duplication: Factor out a protected method to only require setting the `query` variable in both `BM_SortSingleColumnSQL` and `BM_SortMultiColumnSQL` and only pass it to the protected method, where everything else happens.

bakoe committed:
Fix include ordering in sort benchmark

bakoe committed:
Factor out helper method in SQL sort benchmark

jonasnoki commented:
```UnrealScript
Bastian.Koenig@vm-appleton:~/hyrise$ cmake-build-release/hyriseMicroBenchmarks --benchmark_filter=BM_Sort*
2020-01-05 14:56:22
Running cmake-build-release/hyriseMicroBenchmarks
Run on (32 X 2500 MHz CPU s)
CPU Caches:
  L1 Data 32K (x32)
  L1 Instruction 32K (x32)
  L2 Unified 256K (x32)
  L3 Unified 25600K (x32)
Load Average: 0.21, 0.19, 0.08
--------------------------------------------------------------------------------------------
Benchmark                                                  Time             CPU   Iterations
--------------------------------------------------------------------------------------------
SortBenchmark/BM_Sort                               12630939 ns     12624201 ns           56
SortBenchmark/BM_SortSingleColumnSQL                  684324 ns       684231 ns         1026
[PERF] Multiple ORDER BYs are executed one-by-one at src/lib/logical_query_plan/lqp_translator.cpp:272
	Performance can be affected. This warning is only shown once.

SortBenchmark/BM_SortMultiColumnSQL                  1452742 ns      1452732 ns          474
SortPicoBenchmark/BM_SortPico                        1198879 ns      1198870 ns          574
SortSmallBenchmark/BM_SortSmall                       286287 ns       286267 ns         2354
SortLargeBenchmark/BM_SortLarge                       286195 ns       286173 ns         2364
SortReferenceBenchmark/BM_SortReference             16692555 ns     16691590 ns           40
SortReferencePicoBenchmark/BM_SortReferencePico      4326037 ns      4325686 ns          160
SortReferenceSmallBenchmark/BM_SortReferenceSmall     393818 ns       393809 ns         1784
SortReferenceLargeBenchmark/BM_SortReferenceLarge     388630 ns       388612 ns         1786
SortStringBenchmark/BM_SortString                    1921433 ns      1921285 ns          480
SortStringSmallBenchmark/BM_SortStringSmall          1936055 ns      1934903 ns          481
SortStringLargeBenchmark/BM_SortStringLarge          1930920 ns      1927384 ns          480
SortNullBenchmark/BM_SortNullBenchmark                311489 ns       311476 ns         
```
